File: src/storage/councils.ts
import { councilsApi } from '@/api/client'
import {
  normalizeRunState,
  normalizeSocialStructure,
  normalizeSynthesiser,
} from '@/types/council'
import type {
  Council,
  CouncilDeliberation,
  Judge,
  Mediator,
  Seat,
  SeatConfig,
  SocialStructure,
  TokenTotals,
  Turn,
  TurnEvent,
  TurnRunState,
} from '@/types/council'

export interface CouncilSummary {
  id: string
  title: string | null
  createdAt: number
  socialStructure: SocialStructure
  modelIds: string[]
  tokenTotal: TokenTotals
  isDemo?: boolean
}

export interface CreateCouncilInput {
  id: string
  socialStructure: SocialStructure
  seats: Seat[]
  judge?: Judge
  mediator?: Mediator
  deliberation?: CouncilDeliberation
  isDemo?: boolean
}

export interface UpdateSeatInput {
  modelId?: string
  config?: SeatConfig
}

function turnRowToTurn(t: any): Turn {
  const runState = normalizeRunState(t.runState)
  return {
    id: t.id,
    idx: t.idx,
    userMsg: t.userMsg,
    events: t.events,
    tokenTotal: t.tokenTotal,
    ...(t.votingLabels ? { votingLabels: t.votingLabels } : {}),
    ...(t.userImages ? { userImages: t.userImages } : {}),
    ...(runState ? { runState } : {}),
  }
}

function sanitizeDeliberation(d: CouncilDeliberation | undefined): CouncilDeliberation | undefined {
  if (!d) return undefined
  const clean: CouncilDeliberation = {}
  if (d.votingDimensions && d.votingDimensions.length > 0) clean.votingDimensions = d.votingDimensions
  if (d.minCommentLength !== undefined) clean.minCommentLength = d.minCommentLength
  if (d.mediatorMaxRounds !== undefined) clean.mediatorMaxRounds = d.mediatorMaxRounds
  if (d.passDivergence !== undefined) clean.passDivergence = d.passDivergence
  if (d.passPeerAnswers !== undefined) clean.passPeerAnswers = d.passPeerAnswers
  const DELIBERATION_STRING_KEYS = ['participant','votingSystem','votingTemplate','reanswerSystem','reanswerTemplate','judgeTemplate','mediatorTemplate'] as const
  for (const key of DELIBERATION_STRING_KEYS) {
    const v = (d as any)[key]
    if (typeof v === 'string' && v.trim().length > 0) (clean as any)[key] = v
  }
  return Object.keys(clean).length > 0 ? clean : undefined
}

export async function listCouncils(): Promise<CouncilSummary[]> {
  const councils = await councilsApi.list()
  return councils.map(c => ({
    id: c.id,
    title: c.title,
    createdAt: c.createdAt,
    socialStructure: normalizeSocialStructure(c.socialStructure as any),
    modelIds: c.seats.map((s: any) => s.modelId),
    tokenTotal: c.tokenTotal,
    ...(c.isDemo ? { isDemo: true } : {}),
  }))
}

export async function getCouncil(id: string): Promise<Council | null> {
  try {
    const c = await councilsApi.get(id)
    return {
      id: c.id,
      title: c.title,
      createdAt: c.createdAt,
      socialStructure: normalizeSocialStructure(c.socialStructure as any),
      seats: c.seats,
      turns: c.turns.map(turnRowToTurn),
      tokenTotal: c.tokenTotal,
      ...(c.judge ? { judge: normalizeSynthesiser(c.judge) } : {}),
      ...(c.mediator ? { mediator: normalizeSynthesiser(c.mediator) } : {}),
      ...(c.deliberation ? { deliberation: c.deliberation } : {}),
      ...(c.isDemo ? { isDemo: true } : {}),
    }
  } catch {
    return null
  }
}

export async function createCouncil(input: CreateCouncilInput): Promise<Council> {
  try {
    const existing = await getCouncil(input.id)
    if (existing) return existing
  } catch {}
  
  await councilsApi.create({
    ...input,
    deliberation: sanitizeDeliberation(input.deliberation)
  })
  const created = await getCouncil(input.id)
  return created!
}

export async function appendTurn(councilId: string, turn: Turn): Promise<void> {
  const c = await getCouncil(councilId)
  if (!c) throw new Error('council not found')
  const existingTurn = c.turns.find(t => t.id === turn.id)
  if (existingTurn) {
    await councilsApi.updateTurn(councilId, turn.id, turn)
  } else {
    await councilsApi.appendTurn(councilId, turn)
  }
}

export async function patchRunState(councilId: string, turnId: string, patch: Partial<TurnRunState>): Promise<void> {
  const c = await getCouncil(councilId)
  if (!c) return
  const turn = c.turns.find(t => t.id === turnId)
  if (!turn || !turn.runState) return
  await councilsApi.updateTurn(councilId, turnId, {
    ...turn,
    runState: { ...turn.runState, ...patch }
  })
}

export async function clearRunState(councilId: string, turnId: string): Promise<void> {
  const c = await getCouncil(councilId)
  if (!c) return
  const turn = c.turns.find(t => t.id === turnId)
  if (!turn) return
  const updatedTurn = { ...turn }
  delete updatedTurn.runState
  await councilsApi.updateTurn(councilId, turnId, updatedTurn)
}

export async function deleteTurn(councilId: string, turnId: string): Promise<void> {
  await councilsApi.deleteTurn(councilId, turnId)
}

export async function getTurn(councilId: string, turnId: string): Promise<Turn | null> {
  const c = await getCouncil(councilId)
  if (!c) return null
  return c.turns.find(t => t.id === turnId) || null
}

export async function getUnfinishedTurns(turnIds: readonly string[]): Promise<Array<{ councilId: string; turn: Turn }>> {
  if (turnIds.length === 0) return []
  const all = await councilsApi.list()
  const out: Array<{ councilId: string; turn: Turn }> = []
  for (const c of all) {
    const full = await getCouncil(c.id)
    if (full) {
      for (const t of full.turns) {
        if (turnIds.includes(t.id) && t.runState) {
          out.push({ councilId: c.id, turn: t })
        }
      }
    }
  }
  return out
}

export async function replaceEvent(councilId: string, turnId: string, event: TurnEvent): Promise<void> {
  const c = await getCouncil(councilId)
  if (!c) throw new Error('council not found')
  const turn = c.turns.find(t => t.id === turnId)
  if (!turn) throw new Error('turn not found')
  const nextEvents = turn.events.map(e => e.id === event.id ? event : e)
  await councilsApi.updateTurn(councilId, turnId, { ...turn, events: nextEvents })
}

export async function updateSeat(councilId: string, seatId: string, input: UpdateSeatInput): Promise<void> {
  await councilsApi.updateSeat(councilId, seatId, input)
}

export async function addSeat(councilId: string, seat: Seat): Promise<void> {
  await councilsApi.addSeat(councilId, seat)
}

export async function removeSeat(councilId: string, seatId: string): Promise<void> {
  await councilsApi.deleteSeat(councilId, seatId)
}

export async function setJudge(councilId: string, judge: Judge): Promise<void> {
  await councilsApi.update(councilId, { judge })
}

export async function setMediator(councilId: string, mediator: Mediator): Promise<void> {
  await councilsApi.update(councilId, { mediator })
}

export async function setDeliberation(councilId: string, deliberation: CouncilDeliberation): Promise<void> {
  const clean = sanitizeDeliberation(deliberation)
  await councilsApi.update(councilId, { deliberation: clean || null } as any)
}

export async function patchCouncilTitle(councilId: string, title: string): Promise<void> {
  const clamped = title.trim().slice(0, 60)
  await councilsApi.update(councilId, { title: clamped })
}

export async function deleteCouncil(id: string): Promise<void> {
  await councilsApi.delete(id)
}

File: backend/app/routers/councils.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.future import select
from sqlalchemy.orm import selectinload
from app.database import get_db
from app.models import Council, Seat, Turn
from app.schemas import CouncilCreate, CouncilRead, CouncilUpdate, SeatCreate, TurnCreate
from typing import List

router = APIRouter(tags=["councils"])

def serialize_turn(t: Turn) -> dict:
    return {
        "id": t.id, "councilId": t.council_id, "idx": t.idx, "userMsg": t.user_msg, 
        "events": t.events, "tokenTotal": t.token_total, "votingLabels": t.voting_labels, 
        "userImages": t.user_images, "runState": t.run_state
    }

def serialize_seat(s: Seat) -> dict:
    return {
        "id": s.id, "councilId": s.council_id, "modelId": s.model_id, 
        "config": s.config, "pos": s.pos
    }

def serialize_council(c: Council) -> dict:
    return {
        "id": c.id, "title": c.title, "createdAt": c.created_at, "socialStructure": c.social_structure,
        "tokenTotal": c.token_total, "judge": c.judge, "mediator": c.mediator,
        "deliberation": c.deliberation, "isDemo": c.is_demo,
        "seats": [serialize_seat(s) for s in c.seats],
        "turns": [serialize_turn(t) for t in c.turns]
    }

@router.get("", response_model=List[CouncilRead])
async def list_councils(db: AsyncSession = Depends(get_db)):
    result = await db.execute(select(Council).order_by(Council.created_at.desc()).options(selectinload(Council.seats), selectinload(Council.turns)))
    councils = result.scalars().all()
    return [serialize_council(c) for c in councils]

@router.post("", response_model=CouncilRead)
async def create_council(payload: CouncilCreate, db: AsyncSession = Depends(get_db)):
    c = Council(
        id=payload.id, title=payload.title, social_structure=payload.socialStructure,
        judge=payload.judge, mediator=payload.mediator, deliberation=payload.deliberation,
        is_demo=payload.isDemo
    )
    db.add(c)
    for seat in payload.seats:
        s = Seat(id=seat.id, council_id=c.id, model_id=seat.modelId, config=seat.config, pos=seat.pos)
        db.add(s)
    await db.commit()
    result = await db.execute(select(Council).where(Council.id == c.id).options(selectinload(Council.seats), selectinload(Council.turns)))
    return serialize_council(result.scalar_one())

@router.get("/{council_id}", response_model=CouncilRead)
async def get_council(council_id: str, db: AsyncSession = Depends(get_db)):
    result = await db.execute(select(Council).where(Council.id == council_id).options(selectinload(Council.seats), selectinload(Council.turns)))
    c = result.scalar_one_or_none()
    if not c:
        raise HTTPException(status_code=404, detail="Not found")
    return serialize_council(c)

@router.patch("/{council_id}", response_model=CouncilRead)
async def update_council(council_id: str, payload: CouncilUpdate, db: AsyncSession = Depends(get_db)):
    result = await db.execute(select(Council).where(Council.id == council_id))
    c = result.scalar_one_or_none()
    if not c:
        raise HTTPException(status_code=404)
    if payload.title is not None:
        c.title = payload.title
    if payload.judge is not None:
        c.judge = payload.judge
    if payload.mediator is not None:
        c.mediator = payload.mediator
    if payload.deliberation is not None:
        c.deliberation = payload.deliberation
    await db.commit()
    return await get_council(council_id, db)

@router.delete("/{council_id}")
async def delete_council(council_id: str, db: AsyncSession = Depends(get_db)):
    result = await db.execute(select(Council).where(Council.id == council_id))
    c = result.scalar_one_or_none()
    if c:
        await db.delete(c)
        await db.commit()
    return {"status": "ok"}

@router.post("/{council_id}/seats")
async def add_seat(council_id: str, payload: SeatCreate, db: AsyncSession = Depends(get_db)):
    s = Seat(id=payload.id, council_id=council_id, model_id=payload.modelId, config=payload.config, pos=payload.pos)
    db.add(s)
    await db.commit()
    return {"status": "ok"}

@router.patch("/{council_id}/seats/{seat_id}")
async def update_seat(council_id: str, seat_id: str, payload: SeatCreate, db: AsyncSession = Depends(get_db)):
    result = await db.execute(select(Seat).where(Seat.id == seat_id, Seat.council_id == council_id))
    s = result.scalar_one_or_none()
    if s:
        s.model_id = payload.modelId
        s.config = payload.config
        await db.commit()
    return {"status": "ok"}

@router.delete("/{council_id}/seats/{seat_id}")
async def delete_seat(council_id: str, seat_id: str, db: AsyncSession = Depends(get_db)):
    result = await db.execute(select(Seat).where(Seat.id == seat_id, Seat.council_id == council_id))
    s = result.scalar_one_or_none()
    if s:
        await db.delete(s)
        await db.commit()
    return {"status": "ok"}

@router.post("/{council_id}/turns")
async def append_turn(council_id: str, payload: TurnCreate, db: AsyncSession = Depends(get_db)):
    t = Turn(
        id=payload.id, council_id=council_id, idx=payload.idx, user_msg=payload.userMsg,
        events=payload.events, token_total=payload.tokenTotal, voting_labels=payload.votingLabels,
        user_images=payload.userImages, run_state=payload.runState
    )
    db.add(t)
    await db.commit()
    return {"status": "ok"}

@router.patch("/{council_id}/turns/{turn_id}")
async def update_turn(council_id: str, turn_id: str, payload: TurnCreate, db: AsyncSession = Depends(get_db)):
    result = await db.execute(select(Turn).where(Turn.id == turn_id, Turn.council_id == council_id))
    t = result.scalar_one_or_none()
    if t:
        t.events = payload.events
        t.token_total = payload.tokenTotal
        t.run_state = payload.runState
        if payload.votingLabels is not None:
            t.voting_labels = payload.votingLabels
        await db.commit()
    return {"status": "ok"}

@router.delete("/{council_id}/turns/{turn_id}")
async def delete_turn(council_id: str, turn_id: str, db: AsyncSession = Depends(get_db)):
    result = await db.execute(select(Turn).where(Turn.id == turn_id, Turn.council_id == council_id))
    t = result.scalar_one_or_none()
    if t:
        await db.delete(t)
        await db.commit()
    return {"status": "ok"}
